projects/krasnor/src/main.js

```diff
@@ -0,0 +1,574 @@
+        this.internal.LoadingOverlay.insertDefaultStyleToDom();
+        document.body.appendChild(this.internal.LoadingOverlay.domElement);
+
+        this.internal.statisticsPanelLeft = new StatisticsPanel('Subdivided', '0px', '0px', 'none');
+        // this.internal.statisticsPanelLeft.showSubdivisionTime = true;
+        this.internal.statisticsPanelLeft.subdivisionTimeFormat = "hh:mm:ss.ffffff";
+        this.internal.statisticsPanelRight = new StatisticsPanel('Unaltered', '0px', '50%', 'none', true, true);
+        document.body.appendChild(this.internal.statisticsPanelLeft.domElement);
+        document.body.appendChild(this.internal.statisticsPanelRight.domElement);
+
+        // a hidden input field that responsible for loading meshes
+        this.input = document.createElement('input')
+        this.input.setAttribute('type', 'file')
+
+        this.input.addEventListener('change', () => {
+            let file = this.input.files[0]
+            if (!file.name.endsWith('.obj')) {
+                alert('Only .OBJ files are supported')
+            }
+            const r = new FileReader()
+            r.onload = () => this.loadMesh(r.result)
+            r.onerror = () => alert('Cannot import your obj mesh')
+            r.readAsText(file)
+        })
+        document.body.appendChild(this.input)
+
+
+        this.params = {
+            import: () => this.input.click(),
+            export: () => this.exportScreenshot(),
+            downloadMesh: () => this.downloadMesh(),
+            subdivide: () => this.doSubdivide(),
+            showNormals: false,
+            showWireframe: true,
+            flatShading: false,
+            showTexture: true,
+            normalMethod: 'equal-weighted',
+
+            overlayOriginalOverSubdivided: false,
+            statisticsPanelComparisonMethod: this.internal.statisticsPanelLeft.comparisonStyle,
+            statisticsPanelShowTime: this.internal.statisticsPanelLeft.showSubdivisionTime,
+            statisticsPanelSubdivisionTimeFormat: this.internal.statisticsPanelLeft.subdivisionTimeFormat,
+            subdivisions_req: 0.0,
+            boundaryHandling: 'smooth',
+        }
+
+        this.gui = new GUI({hideable: false})
+        this.gui.width = 260; // DEFAULT_WIDTH is 245
+
+        const io = this.gui.addFolder('I/O')
+        io.add(this.params, 'import').name('import mesh')
+        io.add(this.params, 'export').name('export screenshot')
+        io.add(this.params, 'downloadMesh').name('[dbg] export mesh')
+
+        const vis = this.gui.addFolder('Visualization')
+        vis.add(this.params, 'statisticsPanelComparisonMethod', [
+            'none', 'numbers', 'numbers_increase', 'percentage', 'percentage_increase'
+        ]).name('comparison style').onChange(() => {
+                // FIXME BUG Chrome 88 (when Mesh is large ~3Million faces) Even if this callback is empty "() => {}" and nothing would be done, the UI hangs/lags for 1-2 seconds
```

Comment:
This comment is useless as I wrote it days before (immediately after your presentation).

I just got extra time to reproduced your problem in my Chromium 89, and the issue is Major GC blocks the event loop (whereas while subdivision only triggers minor GC).

As you probably already know, V8's major GC does a full mark that scans all the heap-allocated objects, and compact memory (very expensive `memcpy`). The major GC and minor GC are triggered depending on the memory access pattern: when you are actively working on subdivision, the newly allocated vertices and edges are considered active and will never be triggered.

From what I saw in the tracer:

![image](https://user-images.githubusercontent.com/5498964/110220683-83d07300-7ec7-11eb-8b54-da308d7c7327.png)

![image](https://user-images.githubusercontent.com/5498964/110220809-736cc800-7ec8-11eb-941d-8b90afc75035.png)

There are two types of major GC where one spend less time than the other. According to the state of the art implementation of the V8's GC algorithm, the GC phase that remains run on the main thread (will cause STW) contains two parts: finalize of mark, and memory compact:

![image](https://user-images.githubusercontent.com/5498964/110220565-af9f2900-7ec6-11eb-996b-beed5978ef63.png)

This basically explains why the UI blocks: the major GC blocks on the main thread thus the UI freezes. The Major GC may run without compacting memory (shorter one), but may also keep run memory compact (longer one, very costly).

So, this can be considered as a bug of the garbage collector where it constantly moves memory (without any additional memory operation). I never observe this type of behavior in Go since Go's garbage collector is not a moving garbage collector (although I do observe some of the mark phase STW issue in Go) :) But this is a very interesting V8's GC issue to know.

See more about V8's GC: https://v8.dev/blog/trash-talk

Reply:
Wow, thank you very much for the detailed explanation.
This is very interesting.
I didn't know that there are multiple different types of GC in V8.